lib/Monad_Lists.thy

```diff
@@ -0,0 +1,673 @@
+  assumes x: "\<And>x y. \<lbrace>P\<rbrace> m x y \<lbrace>\<lambda>rv. P\<rbrace>"
+  shows      "length xs = length ys \<Longrightarrow> \<lbrace>P\<rbrace> zipWithM_x m xs ys \<lbrace>\<lambda>rv. P\<rbrace>"
+proof (induct xs ys rule: list_induct2)
+  case Nil
+  show ?case
+    by (simp add: zipWithM_x_def sequence_x_def zipWith_def)
+next
+  case (Cons a as b bs)
+  have zipWithM_x_Cons:
+    "\<And>m x xs y ys. zipWithM_x m (x # xs) (y # ys)
+                 = do m x y; zipWithM_x m xs ys od"
+    by (simp add: zipWithM_x_def sequence_x_def zipWith_def)
+  have IH: "\<lbrace>P\<rbrace> zipWithM_x m as bs \<lbrace>\<lambda>rv. P\<rbrace>"
+    by fact
+  show ?case
+    by (simp add: zipWithM_x_Cons) (wp IH x)
+qed
+
+lemma sequence_x_Cons: "\<And>x xs. sequence_x (x # xs) = (x >>= (\<lambda>_. sequence_x xs))"
+  by (simp add: sequence_x_def)
+
+lemma mapM_Cons: "mapM m (x # xs) = (do y \<leftarrow> m x; ys \<leftarrow> (mapM m xs); return (y # ys) od)"
+  by (simp add: mapM_def sequence_def Let_def)
+
+lemma mapM_Nil:
+  "mapM m [] = return []"
+  by (simp add: mapM_def sequence_def)
+
+lemmas mapM_simps = mapM_Nil mapM_Cons
+
+lemma zipWithM_x_mapM:
+ "zipWithM_x f as bs = (mapM (case_prod f) (zip as bs) >>= (\<lambda>_. return ()))"
+  apply (simp add: zipWithM_x_def zipWith_def)
+  apply (induct ("zip as bs"))
+   apply (simp add: sequence_x_def mapM_def sequence_def)
+  apply (simp add: sequence_x_Cons mapM_Cons bind_assoc)
+  done
+
+lemma mapM_x_mapM:
+  "mapM_x m l = (mapM m l >>= (\<lambda>x. return ()))"
+  apply (simp add: mapM_x_def sequence_x_def mapM_def sequence_def)
+  apply (induct l, simp_all add: Let_def bind_assoc)
+  done
+
+lemma mapM_x_Nil:
+  "mapM_x f [] = return ()"
+  unfolding mapM_x_def sequence_x_def
+  by simp
+
+lemma sequence_xappend1:
+  "sequence_x (xs @ [x]) = (sequence_x xs >>= (\<lambda>_. x))"
+  by (induct xs) (simp add: sequence_x_def, simp add: sequence_x_Cons bind_assoc)
+
+lemma mapM_append_single:
+  "mapM_x f (xs @ [y]) = (mapM_x f xs >>= (\<lambda>_. f y))"
+  unfolding mapM_x_def
+  by (simp add: sequence_xappend1)
+
+lemma mapM_x_Cons:
+  "mapM_x m (x # xs) = (do m x; mapM_x m xs od)"
+  by (simp add: mapM_x_def sequence_x_def)
+
+lemma zipWithM_x_mapM_x:
+  "zipWithM_x f as bs = mapM_x (\<lambda>(x, y). f x y) (zip as bs)"
+  apply (subst zipWithM_x_mapM)
+  apply (subst mapM_x_mapM)
+  apply (rule refl)
+  done
+
+lemma zipWithM_x_append1:
+  fixes f :: "'b \<Rightarrow> 'c \<Rightarrow> ('a, unit) nondet_monad"
+  assumes ls: "length xs = length ys"
+  shows "(zipWithM_x f (xs @ [x]) (ys @ [y])) = (zipWithM_x f xs ys >>= (\<lambda>_. f x y))"
+  unfolding zipWithM_x_def zipWith_def
+  by (subst zip_append [OF ls], simp, rule sequence_xappend1)
+
+lemma zipWithM_x_Cons:
+  assumes ls: "length xs = length ys"
+  shows "(zipWithM_x f (x # xs) (y # ys)) = (f x y >>=  (\<lambda>_. zipWithM_x f xs ys))"
+  unfolding zipWithM_x_def zipWith_def
+  by (simp, rule sequence_x_Cons)
+
+
```

Comment:
hmm... if you're using double empty lines to delineate things, consider adding comment/text/section or the careful spacing might not be respected by future people

Reply:
If there was sectioning here before it's pretty much gone by now, so I've removed all of these.